File: dasol/방금 그곡.py
def time_c(t):		# 시간계산
    return int(t.split(':')[0])*60 + int(t.split(':')[1])

def change(x):		# #음 치환
    exc = {'C#':'1','D#':'2', 'F#':'3', 'G#':'4', 'A#':'5'}
    for k, v in exc.items():
        x = x.replace(k, v)
    return x

def solution(m, musicinfos):
    answer = []
    for info in musicinfos:
        info = info.split(',')
        info[3] = change(info[3])
        T = time_c(info[1]) - time_c(info[0])	# 시간계산
        
        if T >= len(info[3]):	# 시간이 길이보다 길면
            M = info[3] * (T//len(info[3])) + info[3][:T%len(info[3])]
        else:					# 시간이 길이보다 짧으면
            M = info[3][:T]
        
        if change(m) in M:		# 들은음이 있으면 /// in 함수를 사용하면 백준에서 시간초과가 나오는 경우가 많았어서 잘 이용안했는데
            answer.append([T, info[2]]) # 이건 반복하는 구조다 보니 매우 빠르게 적용되는 것 같더군요.
        
    if len(answer) == 0:
        return '(None)'
    else:
        return sorted(answer, key=lambda x: -x[0])[0][1] # sorted 메커니즘이 동일한 시간인 경우에 앞선 idx 에 있는 녀석이 그대로 유지 되는 메커니즘인지 확인할 필요가 있어보이네요.

File: younghyo/방금그곡.java
// 프로그래머스 방금 그 곡
//  #음표 소문자로 변경
// 시간 분으로 변환
// 재생 시간 만큼 재생 후 악보생성
// ...
class Solution0207 {
    private String replaceSound(String m) {
        m = m.replaceAll("C#", "c");
        m = m.replaceAll("D#", "d");
        m = m.replaceAll("F#", "f");
        m = m.replaceAll("G#", "g");
        m = m.replaceAll("A#", "a");

        return m;
    }

    private int runningTime(String startInfo, String endInfo) {
        int runningTime = 0;
        int startHour = Integer.parseInt(startInfo.split(":")[0]);
        int startMinute = Integer.parseInt(startInfo.split(":")[1]);
        int endHour = Integer.parseInt(endInfo.split(":")[0]);
        int endMinute = Integer.parseInt(endInfo.split(":")[1]);

        return (endHour - startHour) * 60 + (endMinute - startMinute);
    }


    private String playMusic(String sound, int runningTime) {
        StringBuilder sb = new StringBuilder();
        int soundLength = sound.length();
        for (int i=0; i<runningTime; i++) {
            sb.append(sound.charAt(i % soundLength));
        }
        return sb.toString();
    }

    public String solution(String m, String[] musicinfos) {
        String answer = "(None)";
        m = replaceSound(m);

        int maxRunningTime = 0;
        for (String info : musicinfos) {
            String[] musicInfo = info.split(",");
            int runningTime = runningTime(musicInfo[0], musicInfo[1]);
            String musicName = musicInfo[2];
            String sound = replaceSound(musicInfo[3]);

            String music = playMusic(sound, runningTime);

            if (!music.contains(m)) continue;

            if (runningTime > maxRunningTime) {
                answer = musicName;
                maxRunningTime = runningTime;
            }
        }
        return answer;
    }
}